src/sage/rings/polynomial/polydict.pyx

```diff
@@ -1467,68 +1726,75 @@ cdef class ETuple:
         else:
             return '(' + ', '.join(map(str, self)) + ')'
 
-    def __reduce__(ETuple self):
+    def __reduce__(self):
         """
         EXAMPLES::
 
             sage: from sage.rings.polynomial.polydict import ETuple
             sage: e = ETuple([1,1,0])
-            sage: bool(e == loads(dumps(e)))
+            sage: e == loads(dumps(e))
             True
         """
         cdef size_t ind
-        # this is not particularly fast, but I doubt many people care
-        # if you do, feel free to tweak!
-        d = {self._data[2*ind]: self._data[2*ind+1]
-             for ind from 0 <= ind < self._nonzero}
-        return make_ETuple, (d, int(self._length))
+        d = {self._data[2 * ind]: self._data[2 * ind + 1] for ind in range(self._nonzero)}
+        return ETuple, (d, int(self._length))
 
     # additional methods
 
-    cpdef size_t unweighted_degree(self):
+    cpdef int unweighted_degree(self) except *:
         r"""
         Return the sum of entries.
 
-        ASSUMPTION:
-
-        All entries are non-negative.
-
         EXAMPLES::
 
              sage: from sage.rings.polynomial.polydict import ETuple
-             sage: e = ETuple([1,1,0,2,0])
-             sage: e.unweighted_degree()
+             sage: ETuple([1, 1, 0, 2, 0]).unweighted_degree()
              4
+             sage: ETuple([-1, 1]).unweighted_degree()
+             0
         """
-        cdef size_t degree = 0
+        cdef int degree = 0
         cdef size_t i
-        for i in range(1, 2*self._nonzero, 2):
-            degree += self._data[i]
+        for i in range(self._nonzero):
+            degree += self._data[2 * i + 1]
```

Comment:
Did you check the C code output here? I have lost track if Cython has been able to make these types of loops much more efficient into a simple `for(i=0; i<self._nonzero; i+=2)` or not. It might be good to leave a `TODO` comment here if the C code is still going through Python.

Reply:
Interesting. Cython knows well how to optimize `range(a, b)` (for any kind of `a` and `b` which are first converted to the appropriate integral types). However, it does not optimize `range(a, b, c)`.

Reply:
It is a technical thing with it needing to know the type of `c` is (because of the `i+=c` to `++i` type of things) and they don't really have a good way to do this (even though I think they should really cover some the easier -- or at least trivial -- cases...).